WcaOnRails/app/models/competition.rb

```diff
@@ -851,17 +851,30 @@ def set_wcif_events!(wcif_events)
     ActiveRecord::Base.transaction do
       # Remove extra events.
       self.competition_events.each do |competition_event|
-        competition_event.destroy! unless wcif_events.find { |wcif_event| wcif_event["id"] == competition_event.event.id }
+        wcif_event = wcif_events.find { |e| e["id"] == competition_event.event.id }
+        event_to_be_removed = !wcif_event || !wcif_event["rounds"] || wcif_event["rounds"].empty?
+        if event_to_be_removed
+          raise WcaExceptions::BadApiParameter.new("Cannot remove events from a confirmed competition") if self.isConfirmed?
+          competition_event.destroy!
+        end
       end
 
       # Create missing events.
       wcif_events.each do |wcif_event|
-        competition_events.find_or_create_by!(event_id: wcif_event["id"])
+        event_found = competition_events.find_by_event_id(wcif_event["id"])
+        event_to_be_added = wcif_event["rounds"] && !wcif_event["rounds"].empty?
+        if !event_found && event_to_be_added
+          raise WcaExceptions::BadApiParameter.new("Cannot add events to a confirmed competition") if self.isConfirmed?
+          competition_events.create!(event_id: wcif_event["id"])
+        end
       end
 
       # Update all events.
       wcif_events.each do |wcif_event|
-        competition_events.find_by_event_id!(wcif_event["id"]).load_wcif!(wcif_event)
+        event_to_be_added = wcif_event["rounds"] && !wcif_event["rounds"].empty?
+        if event_to_be_added
+          competition_events.find_by_event_id!(wcif_event["id"]).load_wcif!(wcif_event)
+        end
```

Comment:
Just curious: why changing this behaviour?

Reply:
With the previous version of the code, there were situations where we would create and immediately delete a competition event.

Consider an incoming WCIF that has 3x3x3, but with 0 rounds. In the "# Create missing events." block above, we'd create a competition_event, and then in this "# Update all events" block, we'd call `load_wcif!` for that newly created competition_event, which would then delete the competition_event. This behavior was technically correct, but seemed wasteful to me, so I rewrote it as you see now.